File: frontend/src/api/user-persona-selection.ts
/**
 * Self-service selected-persona API client (Phase 36, PERSONA-03).
 *
 * Uses the shared JWT-bearing `apiClient` (unlike `public-avatar.ts`'s
 * dedicated anonymous `fetch` client) -- every call here requires a real
 * logged-in user; the backend never accepts a client-supplied `user_id`
 * (T-36-21), it always reads `current_user.id` from the JWT dependency.
 */
import apiClient from "./client";

/** Mirrors the backend `SelectedPersonaOut` schema exactly. */
export interface SelectedPersona {
  id: string;
  name: string;
  character: string;
  style: string;
  greeting: string;
}

/** Mirrors the backend `AvatarPersonaOut` schema's fields the switcher needs. */
export interface EnabledPersona {
  id: string;
  name: string;
  character: string;
  style: string;
  greeting: string;
  is_default: boolean;
}

export const userPersonaSelectionApi = {
  /** GET /api/v1/users/me/selected-persona -- always 200, resolves to at
   * least the catalog default. */
  getSelectedPersona: async (): Promise<SelectedPersona> => {
    const { data } = await apiClient.get<SelectedPersona>("/users/me/selected-persona");
    return data;
  },

  /** PUT /api/v1/users/me/selected-persona -- 404 if `personaId` is
   * disabled/unknown, no partial state written. */
  setSelectedPersona: async (personaId: string): Promise<SelectedPersona> => {
    const { data } = await apiClient.put<SelectedPersona>("/users/me/selected-persona", {
      persona_id: personaId,
    });
    return data;
  },

  /** GET /api/v1/personas -- no-auth-required list of enabled personas
   * (Phase 36, PERSONA-01/02); reused here (via the JWT-bearing client, which
   * happily sends the request with or without a token) so the switcher's
   * menu content shares one source of truth with the anonymous persona list. */
  listEnabledPersonas: async (): Promise<EnabledPersona[]> => {
    const { data } = await apiClient.get<EnabledPersona[]>("/personas");
    return data;
  },
};

File: backend/app/api/user_persona_selection.py
"""Self-service selected-persona endpoint (Phase 36, PERSONA-03).

JWT-gated only -- every route here requires a real logged-in user via
`Depends(get_current_user)`; neither route ever accepts or reads a
client-supplied `user_id` (T-36-21). Mounted under `settings.api_prefix`
(unlike the anonymous `/public/avatar/*` surface)."""

from fastapi import APIRouter, Depends
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.services.avatar_persona_service import resolve_active_persona, set_selected_persona

router = APIRouter(prefix="/users/me", tags=["user-persona-selection"])


class SelectedPersonaOut(BaseModel):
    """The caller's resolved active persona -- just enough for the frontend
    switcher trigger + greeting playback (Phase 36, PERSONA-03)."""

    id: str
    name: str
    character: str
    style: str
    greeting: str

    model_config = ConfigDict(from_attributes=True)


class SelectPersonaRequest(BaseModel):
    persona_id: str = Field(..., min_length=1)

    model_config = ConfigDict(from_attributes=False)


@router.get("/selected-persona", response_model=SelectedPersonaOut)
async def get_selected_persona(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> SelectedPersonaOut:
    """Always 200 -- resolves to at least the catalog's default persona via
    `resolve_active_persona()`, never 404."""
    persona = await resolve_active_persona(db, user_id=current_user.id)
    return SelectedPersonaOut.model_validate(persona)


@router.put("/selected-persona", response_model=SelectedPersonaOut)
async def put_selected_persona(
    body: SelectPersonaRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> SelectedPersonaOut:
    """Switch the caller's own active persona (T-36-20: rejects a disabled/
    unknown `persona_id` with 404, no row written -- no partial state)."""
    persona = await set_selected_persona(db, user_id=current_user.id, persona_id=body.persona_id)
    return SelectedPersonaOut.model_validate(persona)
